estransport/connection.go

```diff
@@ -0,0 +1,300 @@
+	connection *Connection
+
+	metrics       *metrics
+	enableMetrics bool
+
+	debugLogger DebuggingLogger
+}
+
+type roundRobinConnectionPool struct {
+	sync.Mutex
+
+	curr int           // Index of the current connection
+	live []*Connection // List of live connections
+	dead []*Connection // List of dead connections
+	orig []*url.URL    // List of original URLs, passed in during initialization
+
+	metrics       *metrics
+	enableMetrics bool
+
+	debugLogger DebuggingLogger
+}
+
+func newSingleConnectionPool(u *url.URL) *singleConnectionPool {
+	cp := singleConnectionPool{connection: &Connection{URL: u}}
+
+	return &cp
+}
+
+func newRoundRobinConnectionPool(u ...*url.URL) *roundRobinConnectionPool {
+	var conns []*Connection
+	for _, url := range u {
+		conns = append(conns, &Connection{URL: url})
+	}
+
+	cp := roundRobinConnectionPool{live: conns, orig: u}
+
+	if cp.enableMetrics {
+		cp.metrics.Lock()
+		cp.metrics.live = cp.live
+		cp.metrics.dead = cp.dead
+		cp.metrics.Unlock()
+	}
+
+	return &cp
+}
+
+// Next returns the connection from pool.
+//
+func (cp *singleConnectionPool) Next() (*Connection, error) {
+	return cp.connection, nil
+}
+
+// Remove is a no-op for single connection pool.
+//
+func (cp *singleConnectionPool) Remove(c *Connection) error {
+	return nil
+}
+
+// Next returns a connection from pool, or an error.
+//
+func (cp *roundRobinConnectionPool) Next() (*Connection, error) {
+	var c *Connection
+
+	cp.Lock()
+	defer cp.Unlock()
+
+	// Try to resurrect a dead connection if no healthy connections are available
+	//
+	if len(cp.live) < 1 {
+		if len(cp.dead) > 0 {
+			if cp.debugLogger != nil {
+				cp.debugLogger.Log("Resurrecting connection ")
+			}
+			c, cp.dead = cp.dead[len(cp.dead)-1], cp.dead[:len(cp.dead)-1] // Pop item
+			c.Lock()
+			if cp.debugLogger != nil {
+				cp.debugLogger.Log(c.URL.String() + "\n")
+			}
+			c.markAsLive()
+			cp.live = append(cp.live, c)
+			c.Unlock()
+
+			return c, nil
+		}
+		return nil, errors.New("no connection available")
+	}
+
+	if cp.curr >= len(cp.live) {
+		cp.curr = len(cp.live) - 1
+	}
+
+	if cp.curr < 0 {
+		return nil, errors.New("no connection available")
+	}
+
+	c = cp.live[cp.curr]
+	cp.curr = (cp.curr + 1) % len(cp.live)
+
+	return c, nil
+}
+
+// Remove removes a connection from the pool.
+//
+func (cp *roundRobinConnectionPool) Remove(c *Connection) error {
+	c.Lock()
+
+	if c.Dead {
+		if cp.debugLogger != nil {
+			cp.debugLogger.Logf("Already removed %s\n", c.URL)
+		}
+		c.Unlock()
+		return nil
+	}
+
+	if cp.debugLogger != nil {
+		cp.debugLogger.Logf("Removing %s...\n", c.URL)
+	}
+	c.markAsDead()
+	cp.scheduleResurrect(c)
+	c.Unlock()
+
+	cp.Lock()
+	defer cp.Unlock()
+
+	// Push item to dead list and sort slice by number of failures
+	cp.dead = append(cp.dead, c)
+	sort.Slice(cp.dead, func(i, j int) bool {
+		c1 := cp.dead[i]
+		c2 := cp.dead[j]
+		c1.Lock()
+		if c1 != c2 {
+			c2.Lock()
+		}
+		defer c1.Unlock()
+		if c1 != c2 {
+			defer c2.Unlock()
+		}
```

Comment:
```go
c1.Lock()
defer c1.Unlock()
if c1 != c2 {
	c2.Lock()
	defer c2.Unlock()
}
```
? (Though I still stand by my statement on Slack that this is unnecessary, and `i != j` is invariant.)

Reply:
I was getting errors when debugging the locking with https://github.com/sasha-s/go-deadlock, and this seemingly unnecessary condition made them go away. I do agree that it's counterintuitive. I'll remove those, and if there's some hidden race condition / deadlock, we'll fix it along the way.

Reply:
Removed in 9be9b94.